### PR TITLE
fix memory leak on nn_bind when error EADDRINUSE returned

### PR DESCRIPTION
**Pull request by bill1600**
Fixes memory leak that happens when nn_bind fails returning EADDRINUSE.

### CODE REVIEW

**Comment by gdamore** [comment]
This is crashing (segfaulting) in the tcp test.

**Comment by bill1600** [comment]
Sorry.  Tests passed in my environment (Fedora 19).

**Comment by bill1600** [comment]
Will resubmit later.

**Comment by bill1600** [comment]
Don't commit this.  I just need to use travis to test.

**Comment by bill1600** [comment]
Tests pass.  This is ready to be merged now.

**Comment by gdamore** [comment]
So, the state machine changes look good, but I'd like to not include the debugging stuff; I realize it was useful to you, the debug stuff needs to be more completely thought out if we're gonna do that, and should be handled as a separate PR.

Ultimately, we would like to get this to just one commit as well.

**Comment by bill1600** [comment]
OK. I will remove the debug.

**Comment by bill1600** [comment]
Removed debug. Should be able to merge now.

File: src/aio/ctx.c [hunk 1]
```diff
@@ -64,9 +64,9 @@ void nn_ctx_leave (struct nn_ctx *self)
     }
 
     /*  Notify the owner that we are leaving the context. */
-    if (nn_fast (self->onleave != NULL))
+    if (nn_fast (self->onleave != NULL)) {
```
Comment (gdamore) [inline src/aio/ctx.c]: This is a style change.  While I personally like having the extra explicit braces, it isn't intrinsic to this change.  We should remove this for now.  (I'm avoiding style changes to the code base, even though I hate the current style.)

File: src/utils/err.c [hunk 2]
```diff
@@ -210,3 +210,4 @@ void nn_win_error (int err, char *buf, size_t bufsize)
 }
 
 #endif
+
```
Comment (gdamore) [inline src/utils/err.c]: Please remove this -- we don't want blank lines at the end of files.

File: tests/tcp.c [hunk 3]
```diff
@@ -26,9 +26,11 @@
 #include "../src/pair.h"
 #include "../src/pubsub.h"
 #include "../src/tcp.h"
+#include "../src/utils/err.h"
```
Comment (gdamore) [inline tests/tcp.c]: This file has only whitespace changes now.  These are not intrinsic to you changes, so lets undo this.

**Comment by gdamore** [comment]
So I've asked for cleanup of some unrelated changes to files -- probably these are things you changed and then changed back as you were trying things out.  We want to boil this to just the change that we want.

Also, I'd like you to submit this as a new PR, with the commits squashed into a single commit, with the commit comment as "Fixes memory leak that happens when nn_bind fails returning EADDRINUSE"  - that way we can see the single issue in one swoop.

(Even better, if we have a github issue already filed for this, the single commit comment would be a line of "fixes #<issue number> <issue synopsis>"   -- see the git commit history for examples of that.

**Comment by bill1600** [comment]
OK, I undertand.
I will clean it up a resubmit as a new PR.

**Comment by bill1600** [comment]
Will resubmit clean.

**Pull request closed without merging**